### PR TITLE
Work around initial login char select issues

### PR DESCRIPTION
**Pull request by Shane7**
A "bunch of bullshit" to get around users having to select a character (that already appears to be selected) to get the play button to light up. The store need to be consolidated (working on that) currently its race condition/synchronization hell. I should be able to flush this code once the store gets sorted out. But I figured a few kludges are worth it now, been seeing login issues/complaints on the forums.

Note: it still has the 30's delay lameness, What is the cost of making this call(patcherAPI.ts line 118):

public getAllChannels() : Channel[] {
    if (this._api.channelData) {
      return Array.prototype.slice.call(this._api.channelData).sort(function(a:Channel, b:Channel) {
        return this.getChannelValue(a) - this.getChannelValue(b);
      }.bind(this));
    }
    return [];
  }

Does that result in a web request or just bounce off the client? How bad would it be to call this every second or so?

### CODE REVIEW

**Comment by codecorsair** [comment]
That call is just reading an array.  There is no talking to the client or a web request, you can call it as often as you like, I would say twice a second should be good.

File: patcher/src/ts/sidebar/ServerSelect.tsx [hunk 1]
```diff
@@ -62,6 +63,13 @@ class SelectServer extends React.Component<SelectServerProps, SelectServerState>
     )
   }
 
+  trySelectCharacter = (shardID:number) => {
+    if(!this.props.charactersState || !this.props.charactersState.characters || 0 == this.props.charactersState.characters.length) return;
+
```
Comment (Shane7) [inline patcher/src/ts/sidebar/ServerSelect.tsx]: Gah, I have a bug here I need to select null before I return.

File: patcher/src/ts/sidebar/ServerSelect.tsx [hunk 2]
```diff
@@ -62,6 +63,13 @@ class SelectServer extends React.Component<SelectServerProps, SelectServerState>
...
+    let filteredCharacters = this.props.charactersState.characters.filter( c => c.shardID == shardID);
```
Comment (Mehuge) [inline patcher/src/ts/sidebar/ServerSelect.tsx]: const not let
=== not ==
https://github.com/airbnb/javascript#comparison--eqeqeq
https://github.com/airbnb/javascript#references--prefer-const

File: patcher/src/ts/sidebar/ServerSelect.tsx [hunk 3]
```diff
@@ -62,6 +63,13 @@ class SelectServer extends React.Component<SelectServerProps, SelectServerState>
...
+    if(filteredCharacters && filteredCharacters.length > 0) this.props.dispatch(selectCharacter(filteredCharacters[0]));
```
Comment (Mehuge) [inline patcher/src/ts/sidebar/ServerSelect.tsx]: if<space>(  //nitpick
https://github.com/airbnb/javascript#whitespace--around-keywords

File: patcher/src/ts/sidebar/ServerSelect.tsx [hunk 4]
```diff
@@ -70,8 +78,8 @@ class SelectServer extends React.Component<SelectServerProps, SelectServerState>
     //is there a required order here?
     dispatch(changeChannel(server.channelInfo));
     dispatch(changeServer(server.serverInfo));
-    dispatch(fetchCharacters());
-    dispatch(selectCharacter(null));
+    if(server && server.shardID) this.trySelectCharacter(server.shardID);
```
Comment (Mehuge) [inline patcher/src/ts/sidebar/ServerSelect.tsx]: if (
https://github.com/airbnb/javascript#whitespace--around-keywords

File: patcher/src/ts/sidebar/ServerSelect.tsx [hunk 5]
```diff
@@ -145,6 +153,11 @@ class SelectServer extends React.Component<SelectServerProps, SelectServerState>
 
     if (!servers && !channels) return null;
 
+    //hack, sort out selected characters/servers/channels here until the underlying store gets sorted out
+    if(!this.props.charactersState.selectedCharacter && this.props.serversState.currentServer) {
```
Comment (Mehuge) [inline patcher/src/ts/sidebar/ServerSelect.tsx]: if (
https://github.com/airbnb/javascript#whitespace--around-keywords

File: patcher/src/ts/sidebar/PatchButton.tsx [hunk 6]
```diff
@@ -20,7 +20,13 @@ import {ServersState} from '../redux/modules/servers';
 import {ChannelState, requestChannels} from '../redux/modules/channels';
 import {CharactersState} from '../redux/modules/characters';
 
-
+//This should probably live elsewhere, maybe it already does?
+enum ChannelID {
+    Hatchery = 4,
```
Comment (Mehuge) [inline patcher/src/ts/sidebar/PatchButton.tsx]: indent wrong

File: patcher/src/ts/sidebar/PatchButton.tsx [hunk 7]
```diff
@@ -259,12 +266,12 @@ class PatchButton extends React.Component<PatchButtonProps, PatchButtonState> {
...
         }
 
-        function isGameChannel(id: number) {
-          return id === 4 || id === 10 || id === 11;
-        }
-
-        if (this.props.charactersState.characters.length == 0 || !this.props.charactersState.selectedCharacter && isGameChannel(this.props.channelsState.selectedChannel.channelID)) {
-          layer1 = <div className='waves-effect btn install-download-btn installing'>Select Character</div>;
+        if (this.props.channelsState.selectedChannel.channelID !== ChannelID.CUBE && !this.props.charactersState.selectedCharacter) {
+            if (this.props.charactersState.characters.length == 0 || this.props.charactersState.characters.filter( c => c.shardID == this.props.serversState.currentServer.shardID).length == 0) {
```
Comment (Mehuge) [inline patcher/src/ts/sidebar/PatchButton.tsx]: === not ==
https://github.com/airbnb/javascript#comparison--eqeqeq

File: patcher/src/ts/sidebar/ServerSelect.tsx [hunk 8]
```diff
@@ -62,6 +63,13 @@ class SelectServer extends React.Component<SelectServerProps, SelectServerState>
     )
   }
 
+  trySelectCharacter = (shardID:number) => {
+    if(!this.props.charactersState || !this.props.charactersState.characters || 0 == this.props.charactersState.characters.length) return;
```
Comment (Mehuge) [inline patcher/src/ts/sidebar/ServerSelect.tsx]: if (
=== not ==
https://github.com/airbnb/javascript#whitespace--around-keywords
https://github.com/airbnb/javascript#comparison--eqeqeq

**Comment by Mehuge** [comment]
Tested this out and it appears to work well.  I couldn't get it to break.

**Pull request merged**